extensions/cornerstone-dynamic-volume/src/panels/PanelGenerateImage.tsx

```diff
@@ -0,0 +1,414 @@
+import {
+  CONSTANTS as cstConstants,
+  Enums as csToolsEnums,
+  segmentation as cstSegmentation,
+  Types as cstTypes,
+  utilities as cstUtils,
+} from '@cornerstonejs/tools';
+
+const DEFAULT_OPTIONS = {
+  TimeFrames: null,
+  Operation: 'SUM',
+};
+
+const SUM = 'SUM';
+const AVG = 'AVERAGE';
+const SUB = 'SUBTRACT';
+
+const operations = [
+  { value: SUM, label: 'SUM', placeHolder: 'SUM' },
+  { value: AVG, label: 'AVERAGE', placeHolder: 'AVERAGE' },
+  { value: SUB, label: 'SUBTRACT', placeHolder: 'SUBTRACT' },
+];
+
+export default function PanelGenerateImage({
+  servicesManager,
+  commandsManager,
+}) {
+  const {
+    viewportGridService,
+    cornerstoneViewportService,
+    displaySetService,
+    hangingProtocolService,
+  } = servicesManager.services;
+  const { t } = useTranslation('PanelGenerateImage');
+  const [generateOptions, setGenerateOptions] = useState(DEFAULT_OPTIONS);
+  const [timeOptions, setTimeOptions] = useState([]);
+  const [rangeValues, setRangeValues] = useState([]);
+  const [sliderValues, setSliderValues] = useState([]);
+  const [timeFramesToUse, setTimeFramesToUse] = useState([]);
+  const [computedDisplaySet, setComputedDisplaySet] = useState(null);
+  const [myDynamicVolume, setMyDynamicVolume] = useState(null);
+  const [myComputedVolume, setMyComputedVolume] = useState(null);
+  const [frameRate, setFrameRate] = useState(20);
+  const [uuidComputedVolume, setUuidComputedVolume] = useState(
+    csUtils.uuidv4()
+  );
+
+  // Cine states
+  // cineState is passed from the reducer in cine provider, cineService is the
+  // api passed from the reducer
+  const [cineState, cineService] = useCine();
+  const { inCineEnabled, cines } = cineState;
+
+  const handleGenerateOptionsChange = generateOptions => {
+    setGenerateOptions(prevState => {
+      const newState = { ...prevState };
+      Object.keys(generateOptions).forEach(key => {
+        if (typeof generateOptions[key] === 'object') {
+          newState[key] = {
+            ...prevState[key],
+            ...generateOptions[key],
+          };
+        } else {
+          newState[key] = generateOptions[key];
+        }
+      });
+      return newState;
+    });
+  };
+
+  // Establish a reference to the viewer API context
+  const { activeViewportIndex, viewports } = viewportGridService.getState();
+  const displaySetInstanceUID = viewports[0].displaySetInstanceUIDs[0];
+  const volumeLoaderScheme = 'cornerstoneStreamingDynamicImageVolume'; // Loader id which defines which volume loader to use
+  const computedVolumeId = `cornerstoneStreamingImageVolume:${uuidComputedVolume}`;
+  const dynamicVolumeId = `${volumeLoaderScheme}:${displaySetInstanceUID}`;
+
+  useEffect(() => {
+    // ~~ Subscription
+    const added = cornerstoneViewportService.EVENTS.VIEWPORT_DATA_CHANGED;
+    const subscriptions = [];
+
+    // Used to make sure initialization of computed volume only runs once
+    let initStatus = false;
+
+    [added].forEach(evt => {
```

Comment:
if you have only one event, just use that, you don't need to create array and foreach